DataStructures/Lists/CircularQueue.cs

```diff
@@ -0,0 +1,227 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+
+namespace DataStructures.CircularQueueSpace
+{
+    [Serializable]
+    public class CircularQueue<T>:ICollection<T>,IEnumerable<T>
+    {
+        private int? _head;       // remove from head
+        private int? _tail;       // insert at tail
+
+        /// <summary>
+        /// Creates a non readonly Queue
+        /// </summary>
+        public CircularQueue()
+        {
+            Init(false);
+            QueueList = new List<T>();
+        }
+
+        /// <summary>
+        /// Creates a Queue
+        /// </summary>
+        /// <param name="isReadOnly">Determines if the Queue is readonly</param>
+        public CircularQueue(bool isReadOnly)
+        {
+            Init(isReadOnly);
+            QueueList = new List<T>();
+        }
+
+        /// <summary>
+        /// Creates a Queue
+        /// </summary>
+        /// <param name="quesize">Capacity of the Queue</param>
+        /// <param name="isReadOnly">Determines if the Queue is readonly</param>
+        public CircularQueue(int quesize, bool isReadOnly)
+        {
+            Init(isReadOnly);
+            QueueList = new List<T>(quesize);
+        }
+
+        /// <summary>
+        /// Creates a Queue
+        /// </summary>
+        /// <param name="quesize">Capacity of the Queue</param>
+        public CircularQueue(int quesize)
+        {
+            Init(false);
+            QueueList = new List<T>(quesize);
+        }
+
+        /// <summary>
+        /// Init the Queue
+        /// </summary>
+        /// <param name="isReadOnly">Set if the Queue is going to be readonly</param>
+        public void Init(bool isReadOnly)
```

Comment:
Why is `Init()` a public method? I think it should be private as there is already a reset method - `Clear()`.